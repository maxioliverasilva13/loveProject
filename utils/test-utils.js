import { TodosService } from "@/services/TodosService";
import GlobalSlice from "@/slices/GlobalSlice";
import { configureStore } from "@reduxjs/toolkit";
import { render as rtlRender } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";

function render(
  ui,
  {
    preloadedState,
    store = configureStore({
      reducer: {
        GlobalSlice,
        [TodosService.reducerPath]: TodosService.reducer,
      },
      preloadedState,
      middleware: getDefaultMiddleware => getDefaultMiddleware().concat(TodosService.middleware),
    }),

    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from "@testing-library/react";
// override render method
export { render };
