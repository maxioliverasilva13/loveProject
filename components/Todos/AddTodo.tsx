import Button from "@/components/shared/Button/Button";
import Loading from "@/components/shared/Loading/Loading";
import { useAddTodoMutation } from "@/services/TodosService";
import { useGlobal } from "@/slices/GlobalSlice";
import { classNames } from "@/utils/classNames";
import { useCallback, useState, KeyboardEvent } from "react";

const AddTodo = () => {
  const [isActive, setIsActive] = useState(false);
  const autoFocus = useCallback((el: HTMLInputElement): void | null => (el ? el.focus() : null), []);
  const [addTodo, { isLoading: isAdding }] = useAddTodoMutation();
  const { darkMode } = useGlobal();

  const handleKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
    if (e?.code === "Enter") {
      addTodo({ title: e.currentTarget?.value, completed: false });
      setIsActive(false);
    }
  };

  return (
    <div className="text-center text-lg cursor-pointer mt-4">
      {isActive ? (
        <input
          className={classNames(
            "w-full bg-transparent rounded border border-1 border-gray-300 outline-0 px-4 py-2 text-gray-400",
            darkMode ? "text-white" : "text-black",
          )}
          ref={autoFocus}
          onKeyDown={handleKeyDown}
          onBlur={() => setIsActive(false)}
          type="text"
        />
      ) : (
        <>
          {isAdding && <Loading />}
          <Button label="Add a new Todo" variant="info" onClick={() => setIsActive(!isActive)} />
        </>
      )}
    </div>
  );
};

export default AddTodo;
