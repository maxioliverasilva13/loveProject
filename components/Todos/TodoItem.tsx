import Loading from "@/components/shared/Loading/Loading";
import { useDeleteTodoMutation, useUpdateTodoMutation } from "@/services/TodosService";
import { Todo } from "@/types/Todo";
import { classNames } from "@/utils/classNames";
import { useTheme } from "next-themes";
import { useEffect, useState } from "react";
import { Check, Trash } from "react-feather";
import Swal from "sweetalert2";

interface Props {
  item: Todo;
}

const TodoItem = ({ item = { title: "", completed: false } }: Props) => {
  const [updateTodo, { isLoading: isUpdating, isError: errorUpdating }] = useUpdateTodoMutation();
  const [deleteTodo, { isLoading: isDeleting }] = useDeleteTodoMutation();
  const [isCompleted, setIsCompleted] = useState(item.completed);
  const { theme } = useTheme();
  const isDarkMode = theme === "dark";

  function handleToggle() {
    if (isUpdating || isDeleting) return;

    setIsCompleted(!item.completed);
    updateTodo({ ...item, completed: !item.completed });
  }

  const handleDeleteTodo = async (item: Todo) => {
    await deleteTodo(item);
  };

  useEffect(() => {
    if (errorUpdating) {
      setIsCompleted(!item.completed);
    }
  }, [errorUpdating, item.completed]);

  return (
    <div
      className={classNames(
        "grid grid-cols-[30px_auto_30px] items-center gap-1 rounded p-4 mt-1 bg-white border transition-all duration-75",
        isDarkMode ? "border-gray-700 text-black" : "border-gray-300",
        isCompleted ? "line-through opacity-70" : "decoration-0",
      )}
      data-testid={item.title}
    >
      <button
        className="w-[30px] h-[30px] rounded-full border border-gray-700 cursor-pointer pt-1 pl-1 transition-all duration-300"
        role="checkbox"
        aria-checked={isCompleted}
        onClick={handleToggle}
      >
        {isCompleted && <Check size="22px" />}
      </button>
      <h4 className="text-lg m-0 p-0 overflow-hidden overflow-ellipsis whitespace-nowrap">{item.title}</h4>
      {isDeleting ? (
        <Loading />
      ) : (
        <Trash
          data-testid={`deleteBtn${item.title}`}
          onClick={() => {
            Swal.fire({
              title: "Are you sure?",
              icon: "warning",
              showCancelButton: true,
              confirmButtonColor: "#3085d6",
              cancelButtonColor: "#d33",
              confirmButtonText: "Yes, delete it!",
            }).then(result => {
              if (result.isConfirmed) {
                handleDeleteTodo(item);
              }
            });
          }}
          color="red"
        />
      )}
    </div>
  );
};

export default TodoItem;
