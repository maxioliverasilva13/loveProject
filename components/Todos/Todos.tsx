import AddTodo from "./AddTodo";
import TodoItem from "./TodoItem";
import { Todo } from "@/types/Todo";

interface Props {
  todos: Todo[];
}

const Todos = ({ todos = [] }: Props) => {
  return (
    <div className="w-full">
      {todos.map(todo => (
        <TodoItem key={todo._id} item={todo} />
      ))}
      <AddTodo />
    </div>
  );
};

export default Todos;
