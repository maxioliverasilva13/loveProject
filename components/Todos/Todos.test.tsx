import Todos from "./Todos";
import { render, fireEvent, cleanup } from "@/utils/test-utils";
import { RenderResult, act } from "@testing-library/react";
import React from "react";

window.scrollTo = jest.fn();

const data = [
  { _id: "1", title: "Test Item 1", completed: false },
  { _id: "2", title: "Test Item 2", completed: false },
];

describe("Todos", () => {
  let Component: RenderResult;

  beforeEach(async () => {
    act(() => {
      Component = render(<Todos todos={data} />);
    });
  });

  afterAll(() => cleanup());

  it("should render the list of items", () => {
    expect(Component.container).toBeTruthy();
    expect(Component.getByText("Test Item 1")).toBeInTheDocument();
    expect(Component.getByText("Test Item 2")).toBeInTheDocument();

    //snapshot
    expect(Component.asFragment()).toMatchSnapshot();
  });

  it("should update on complete", async () => {
    expect(Component.getAllByRole("checkbox", { checked: false })[0]).toBeInTheDocument();

    // await act(async () => (Component = render(<Todos todos={data} />)));
    await act(async () => fireEvent.click(Component.getAllByRole("checkbox")[0]));
    expect(Component.getAllByRole("checkbox", { checked: true })[0]).toBeInTheDocument();
  });

  it("should enable input when clicking add todo", () => {
    {
      expect(Component.queryByRole("textbox")).not.toBeInTheDocument();
      fireEvent.click(Component.getByText("Add a new Todo"));
      expect(Component.getByRole("textbox")).toBeInTheDocument();
    }
  });
});
