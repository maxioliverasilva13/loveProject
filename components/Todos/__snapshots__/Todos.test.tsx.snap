// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos should render the list of items 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <div
      class="grid grid-cols-[30px_auto_30px] items-center gap-1 rounded p-4 mt-1 bg-white border transition-all duration-75 border-gray-300 decoration-0"
      data-testid="Test Item 1"
    >
      <button
        aria-checked="false"
        class="w-[30px] h-[30px] rounded-full border border-gray-700 cursor-pointer pt-1 pl-1 transition-all duration-300"
        role="checkbox"
      />
      <h4
        class="text-lg m-0 p-0 overflow-hidden overflow-ellipsis whitespace-nowrap"
      >
        Test Item 1
      </h4>
      <svg
        data-testid="deleteBtnTest Item 1"
        fill="none"
        height="24"
        stroke="red"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="3 6 5 6 21 6"
        />
        <path
          d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
        />
      </svg>
    </div>
    <div
      class="grid grid-cols-[30px_auto_30px] items-center gap-1 rounded p-4 mt-1 bg-white border transition-all duration-75 border-gray-300 decoration-0"
      data-testid="Test Item 2"
    >
      <button
        aria-checked="false"
        class="w-[30px] h-[30px] rounded-full border border-gray-700 cursor-pointer pt-1 pl-1 transition-all duration-300"
        role="checkbox"
      />
      <h4
        class="text-lg m-0 p-0 overflow-hidden overflow-ellipsis whitespace-nowrap"
      >
        Test Item 2
      </h4>
      <svg
        data-testid="deleteBtnTest Item 2"
        fill="none"
        height="24"
        stroke="red"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="3 6 5 6 21 6"
        />
        <path
          d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
        />
      </svg>
    </div>
    <div
      class="text-center text-lg cursor-pointer mt-4"
    >
      <button
        class="px-2 py-2 rounded cursor-pointer transition-all text-white bg-gray-400 hover:bg-gray-500"
      >
        Add a new Todo
      </button>
    </div>
  </div>
</DocumentFragment>
`;
